lib/ecto/association.ex

```diff
@@ -189,8 +189,9 @@ defmodule Ecto.Association do
     rewrite_ix = assoc.ix
     [assoc|joins] = Enum.map([assoc|joins], &rewrite_join(&1, rewrite_ix))
 
+    {_, schema} = assoc.source
     %{query | wheres: [assoc_to_where(assoc)|query.wheres], joins: joins,
-              from: merge_from(query.from, assoc.source), sources: nil}
+              from: merge_from(query.from, assoc.source), sources: nil, prefix: schema.__schema__(:prefix)}
```

Comment:
I think the prefix should be set here: https://github.com/shhavel/ecto/blob/51e69a35b036ebf3efc4abe625b8d829f86b5011/lib/ecto/association.ex#L163

Only on the right side of `||`. If a query is given, we should likely respect the prefix in the query because we are going to respect the from in that query too.